### PR TITLE
feat: tests for metrics

### PR DESCRIPTION
**Pull request by pboros**
Have a way to test if a certain metric is appearing in the actual metrics output based on the status json output.

### CODE REVIEW

**Review by terbepetra — APPROVED** [review]

File: metrics/group_data_test.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// limitations under the License.
+
+package metrics
+
+import (
+	"testing"
+)
+
+func TestNewDataMetricGroup(t *testing.T) {
+	metrics := getMetricsFromTestFile(t, "status-single-basic.json")
```
Comment (rbarabas) [inline metrics/group_data_test.go:22]: Is this file supposed to be checked in?

Comment (rbarabas) [reply metrics/group_data_test.go:22]: Was the file added prior?

Comment (pboros) [reply metrics/group_data_test.go:22]: Yes, it was added prior, other tests are using it.

File: metrics/metricinfo.go [hunk 2]
```diff
@@ -96,6 +112,85 @@ func (m *MetricInfo) collectOnce(ctx context.Context) error {
 	return nil
 }
 
+func (m *MetricInfo) collectOnceFromFile(t *testing.T, fileName string) {
```
Comment (himank) [inline metrics/metricinfo.go:115]: Can these functions simply return an error and then the caller can check for errors?

File: metrics/metricinfo.go [hunk 3]
```diff
@@ -96,6 +112,85 @@ func (m *MetricInfo) collectOnce(ctx context.Context) error {
...
+	assert.Nil(t, err, "error parsing metric line with regexp")
+	match := fullMetricRe.FindStringSubmatch(line)
+	if len(match) == 4 {
+		return fetchedMetric{metricKey: match[1], tags: match[2], metricValue: match[3]}
+	} else {
+		return fetchedMetric{}
+	}
+}
+
+func checkMetrics(t *testing.T, metrics []fetchedMetric, expectedMetrics []string) {
```
Comment (himank) [inline metrics/metricinfo.go:179]: same for all the methods ^

File: metrics/metricinfo.go (hunk 2 above)
Comment (pboros) [reply metrics/metricinfo.go]: Refactored it like that.

File: metrics/metricinfo.go (hunk 3 above)
Comment (pboros) [reply metrics/metricinfo.go]: Refactored it.

**Review by himank — APPROVED** [review]

File: metrics/group_workload_keys.go [hunk 4]
```diff
@@ -26,5 +29,9 @@ func NewWorkloadKeysMetricGroup(info *MetricInfo) *WorkloadKeysMetricGroup {
 
 func (w *WorkloadKeysMetricGroup) GetMetrics(status *models.FullStatus) {
 	scope := w.GetScopeOrExit("default")
+	if status == nil || status.Cluster == nil || status.Cluster.Workload == nil || status.Cluster.Workload.Keys == nil || &status.Cluster.Workload.Keys.Read == nil {
```
Comment (himank) [inline metrics/group_workload_keys.go:32]: nit: this can be moved to a helper method `IsValidStatus()`

Comment (pboros) [reply metrics/group_workload_keys.go:32]: I will move the high level ones in the next pr, whole sections can be missing here.

**Pull request merged**